### ISSUE TITLE
[STORM-2006] Storm metrics feature improvement: support per-worker level metrics aggregation

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#5788]
<p>Storm provides per-task level metrics which could be huge when topology has a number of tasks. <br/>
Task level metric is useful for determining load balance between tasks, but it doesn't need to be time-series fashion.</p>

<p>Before introducing topology level component like TopologyMaster for JStorm, we can utilize SystemBolt to aggregate task level metrics to per-worker level metrics.</p>

<p>We should provide options and this feature should be turned off by default to keep backward compatibility. </p>

---
<details><summary><i>Originally reported by <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-2006" target="_blank">Storm metrics feature improvement: support per-worker level metrics aggregation</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Invalid
<li><b>resolved</b>: 2016-11-15T03:47:56+00:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user HeartSaVioR opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1594" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1594</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a> Storm metrics feature improvement: support per-worker level metrics aggregation (1.x)</p>

<ul>
	<li>SystemBolt handles task level metrics via two mode
	<ul>
		<li>non-aggregate: same to previous, just applying expansion and pass to MetricConsumerBolts</li>
		<li>aggregate: apply expansion, do aggregation, pass aggregated metrics to MetricConsumerBolts</li>
	</ul>
	</li>
	<li>all task level metrics should pass by SystemBolt within its worker
	<ul>
		<li>it drops all connections between tasks and MetricsConsumerBolts</li>
		<li>only SystemBolts and MetricConsumerBolts will be connected</li>
	</ul>
	</li>
	<li>move configurations: expandMapType and metricNameSeparator to global
	<ul>
		<li>since SystemBolt needs to handle expansion when both worker level aggregation and expandMapType are turned on</li>
	</ul>
	</li>
</ul>


<p>    Manually tested via matrix of options : aggregate x expand</p>

<p>    I'm not sure how to create unit tests playing with SystemBolt, but will try to come up unit tests.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/HeartSaVioR/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/HeartSaVioR/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a>-1.x</p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1594.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1594.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1594</p>

<hr />
<p>commit b2acd027768c30b71357a55a8eccd13a485089f3<br/>
Author: Jungtaek Lim <kabhwan@gmail.com><br/>
Date:   2016-07-25T13:43:02Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a> Storm metrics feature improvement: support per-worker level metrics aggregation</p>

<ul>
	<li>SystemBolt handles task level metrics via two mode
	<ul>
		<li>non-aggregate: same to previous, just applying expansion and pass to MetricConsumerBolts</li>
		<li>aggregate: apply expansion, do aggregation, pass aggregated metrics to MetricConsumerBolts</li>
	</ul>
	</li>
	<li>all task level metrics should pass by SystemBolt within its worker
	<ul>
		<li>it drops all connections between tasks and MetricsConsumerBolts</li>
		<li>only SystemBolts and MetricConsumerBolts will be connected</li>
	</ul>
	</li>
	<li>move configurations: expandMapType and metricNameSeparator to global
	<ul>
		<li>since SystemBolt needs to handle expansion when both worker level aggregation and expandMapType are turned on</li>
	</ul>
	</li>
</ul>


<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>GitHub user HeartSaVioR opened a pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a> Storm metrics feature improvement: support per-worker level metrics aggregation</p>

<p>    PR for 1.x : #1594 </p>

<ul>
	<li>SystemBolt handles task level metrics via two mode</li>
	<li>non-aggregate: same to previous, just applying expansion and pass to MetricConsumerBolts</li>
	<li>aggregate: apply expansion, do aggregation, pass aggregated metrics to MetricConsumerBolts</li>
	<li>all task level metrics should pass by SystemBolt within its worker</li>
	<li>it drops all connections between tasks and MetricsConsumerBolts</li>
	<li>only SystemBolts and MetricConsumerBolts will be connected</li>
	<li>move configurations: expandMapType and metricNameSeparator to global</li>
	<li>since SystemBolt needs to handle expansion when both worker level aggregation and expandMapType are turned on</li>
</ul>


<p>    This could break #1445 and it's already broken. We ideally want to stop adding new feature on 1.x and concentrate on 2.x but life is not easy.<br/>
    I can volunteer to fix the conflict if this PR breaks #1445 in many places.</p>

<p>You can merge this pull request into a Git repository by running:</p>

<p>    $ git pull <a href="https://github.com/HeartSaVioR/storm" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/HeartSaVioR/storm</a> <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a></p>

<p>Alternatively you can review and apply these changes as the patch at:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595.patch" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595.patch</a></p>

<p>To close this pull request, make a commit to your master/trunk branch<br/>
with (at least) the following in the commit message:</p>

<p>    This closes #1595</p>

<hr />
<p>commit 8d1bc0b7ca2e50d45af3bc1bf41944925611910e<br/>
Author: Jungtaek Lim <kabhwan@gmail.com><br/>
Date:   2016-07-25T13:43:02Z</p>

<p>    <a href="https://issues.apache.org/jira/browse/STORM-2006" title="Storm metrics feature improvement: support per-worker level metrics aggregation" class="issue-link" data-issue-key="STORM-2006"><del>STORM-2006</del></a> Storm metrics feature improvement: support per-worker level metrics aggregation</p>

<ul>
	<li>SystemBolt handles task level metrics via two mode
	<ul>
		<li>non-aggregate: same to previous, just applying expansion and pass to MetricConsumerBolts</li>
		<li>aggregate: apply expansion, do aggregation, pass aggregated metrics to MetricConsumerBolts</li>
	</ul>
	</li>
	<li>all task level metrics should pass by SystemBolt within its worker
	<ul>
		<li>it drops all connections between tasks and MetricsConsumerBolts</li>
		<li>only SystemBolts and MetricConsumerBolts will be connected</li>
	</ul>
	</li>
	<li>move configurations: expandMapType and metricNameSeparator to global
	<ul>
		<li>since SystemBolt needs to handle expansion when both worker level aggregation and expandMapType are turned on</li>
	</ul>
	</li>
</ul>


<hr />

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72313373" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72313373</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    we should make this as true and document it. As keeping this false can easily put lot of metrics into users metric storage. If they are not aware of it can get into issues.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    +1 . Minor nit.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72352692" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72352692</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @harshach <br/>
    Set this to 'true' breaks current metrics consumer implementations, since it delegates aggregation to metrics consumer by passing list of values. We need to guide the way how to aggregate these values to well-known metrics consumer implementations.</p>

<p>    Btw, we should try avoiding to address current metrics feature and start re-designing new metrics feature. To tell the truth, this feature is actually closer to a hot-fix instead of improvement. <br/>
    Even I found the design issue while implementing this. Current metrics feature doesn't allow us to aggregate the values, since we can't add the metadata (which function we need to apply to aggregate) to metrics. Note that we expand metrics from value of metric, so adding metadata to metric doesn't make sense. Metric shouldn't use Object as value, especially some types are intended to be expanded.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72355902" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72355902</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    I also think current pluggable metrics consumer design is too flexible. Users rely on metrics consumers which are open source from individual or specific vendor, which is easy to be not maintained. I guess most popular implementation of metrics consumer is verisign/storm-graphite, and it's not maintained for months. We should provide stable consumer implementation. If we don't want to handle various time-series DBs, we can open at most pluggable connector interface.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    +1</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72479602" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72479602</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @HeartSaVioR <br/>
    If I'm following the code/conversation correctly, this flag essentially changes the IMetricConsumer interface (specifically the value set in the DataPoint class - which is of type Object).  That makes me very uneasy.</p>

<p>    Is there a way to make the interface more strongly typed?</p>

<p>    Thinking out loud: What about splitting the interface in two: One that does aggregation and one that doesn't, then switching calls based on which interface is implemented?</p>

<p>    I'm not opposed to breaking backwards compatibility and bumping up the major version if necessary. In other words I'd rather see a clean design than preserve backwards compatibility with respect to the metrics features.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72504045" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72504045</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @ptgoetz I don't see any interface changes. Its just additional config in absence of it or the default it will continue to work with older implementation of IMetricsConsumer.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72526172" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72526172</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @ptgoetz <br/>
    As you said this is just due to a design flaw of current metrics feature. <br/>
    Return type of getValueAndReset (Object) just makes whole things opaque, which is really bad.<br/>
    Also we even didn't document a <b>implicit contract</b> that Map type of value of metric should be expanded to multiple metrics. How to process metrics is completely up to metrics consumer developer, and some metrics consumers even don't expand Map type of value. I'm introducing similar thing from this change, which I think it is still bad, but if you read the conversation, you may notice that there's no other choice with current metrics.</p>

<p>    @harshach @ptgoetz <br/>
    This must be a last call of new feature/improvement of current metrics feature. To tell the truth, I should never improve current metrics feature and concentrate on new metrics feature, but I didn't and couldn't, because of several reasons. My bad.</p>

<p>    Actually my bad decision started from our progress of Storm 2.0.0 plan. For the first time I'm just waiting for JStorm merger phase 2 so that JStorm metrics feature to be applied. But we even didn't finish phase 1 in 6 months, which is being said that it is expected to be finished in 1Q. So I was end up with going on my own work, but have been considered about the backward compatibility, and once I started considering that, I couldn't get out of it.</p>

<p>    I'm still in doubt that we can finish JStorm merger by end of this year, since phase 1 seems to be no progress on more than 1 month. If we would want to concentrate port & merger works and make this finally happen, I can wait for that. (Sure I'll participate the works.) But if it's out of our mind, we should have the time to design and implement new metrics feature.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72539748" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72539748</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @HeartSaVioR Just to be clear, please don't take my comments to be a criticism of your work. You've done a fantastic job with metrics improvements.</p>

<p>    Through that work it's become apparent that the existing API for metrics consumers is lacking in a number of ways. I view this as an opportunity to correct that.</p>

<p>    We're incrementally moving toward semantic versioning. I'm not opposed to breaking backward compatibility with the existing metrics API and moving this branch to 2.x. The fact that master is "2.0" is just a technicality that can easily be changed.</p>

<p>    I'd like to fix the API to better reflect the work you've done, rather than to implement a one-off fix.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72548331" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72548331</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    @ptgoetz No worries. I didn't take your comments in that way. It was just my criticism of *<b>current metrics API</b>*.</p>

<p>    Storm community discussed about metrics several times, and the only one that I can get is the needs of new metrics feature. </p>

<p>    This is a last discussion which also contains requirements of metrics feature. I don't think we should address all of them, but addressing much of them would be great.<br/>
    <a href="http://mail-archives.us.apache.org/mod_mbox/storm-dev/201605.mbox/%3CCAF5108jB=4aL0Z1qMMnEe7U4Yx_9TZjD11BJYhKtkTtVJoTdhQ@mail.gmail.com%3E" class="external-link" target="_blank" rel="nofollow noopener">http://mail-archives.us.apache.org/mod_mbox/storm-dev/201605.mbox/%3CCAF5108jB=4aL0Z1qMMnEe7U4Yx_9TZjD11BJYhKtkTtVJoTdhQ@mail.gmail.com%3E</a></p>

<p>    Adding fixes to current metrics seems not competitive to recent metrics from others stream frameworks, and even JStorm, which is the thing we would eventually evaluate (I already did a first pass and looks promising) and port if it's better. This patch only has a value with current metrics, and I don't want to break backward compatibility because of will-be-deprecated-and-dropped feature.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72548650" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72548650</a></p>

<p>    &#8212; Diff: conf/defaults.yaml &#8212;<br/>
    @@ -259,6 +259,10 @@ topology.disruptor.batch.size: 100<br/>
     topology.disruptor.batch.timeout.millis: 1<br/>
     topology.disable.loadaware: false<br/>
     topology.state.checkpoint.interval.ms: 1000<br/>
    +topology.metrics.aggregate.per.worker: false<br/>
    &#8212; End diff &#8211;</p>

<p>    FYI, I tried to port JStorm metrics to 1.x some months ago, but it touches many paths of codebase so we would end up having hard time to do the work if port is not finished.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @HeartSaVioR  You improved the metrics collection and quality of those metrics as well. This is immensely useful in debugging any topologies and operationalizing the Storm it self. I am not sure if there any users (atleast I haven't seen one) using the existing topology.metrics.consumer as its performance is bad and can get the topology to a halt.<br/>
     In light of these issues what you did here is will be extremely helpful and learnings from here will help in the 2.0 changes as well. <br/>
    Having said that this fix critical as we end up flooding the storage of metrics with exponential number of metrics.<br/>
    @ptgoetz "that contract is broken when this flag is set to true".  Yes and we are shipping this as false. So Unless user is consciously makes this as true its not breaking anything. Sufficient documentation around this will help on what to expect.<br/>
      We also need to make a judgement call on what it will impact . With this fix in users can go ahead store the metrics much more optimal fashion than what was happening before.<br/>
    IMHO this is not a backward-incompatible change and it will help users in storing the metrics better.</p>

<p>    I am still +1 on merging this patch.</p>

<p>    "Btw, we should try avoiding to address current metrics feature and start re-designing new metrics feature. To tell the truth, this feature is actually closer to a hot-fix instead of improvement."</p>

<p>    @HeartSaVioR  agree. But this will get the users to start using the metrics feature in storm. Before any of these changes all users depends on are UI metrics and with these changes in they can more accurately look at whats going on the cluster. Also they can add metrics into their topology. we definitely should improve this feature for 2.0.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @harshach @ptgoetz <br/>
    I believe the best approach is letting users don't care about them.</p>

<p>    While I still think we need to renew metrics feature, I also think one of valid approach for current metrics is spending some more time to provide some official implementations of metrics consumers which covers popular time-series DBs like Grafana, and OpenTSDB, InfluxDB, and some others if we have ideas.</p>

<p>    This makes us feeling similar to dog-fooding, and will show how good or how bad is the metrics feature transparently.<br/>
    (To tell the truth, I realized the limitations of current metrics feature clearly while playing with Ambari Metric Service Storm sink.)<br/>
    Contracts will be mostly on our own, so no delegation for the hard thing on users or dev. of metric consumers.</p>

<p>    Just I'm not 100% sure open source metrics consumer projects will reflect our changed metrics API or contract in time. So while we're improving the feature, it could unlikely get to the end-users. This is my another concern.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    cc. @abhishekagarwal87 <br/>
    I guess you seems to operate Storm metrics in production, so would like to hear your thought about this. (Not meant to review the code. I don't have operation experience for metrics on production, so need some help on this area.)</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user ptgoetz commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @HeartSaVioR I don't think I was very clear in articulating my proposal. What I'm saying is rather than making this a configuration switch, make it dynamic based on what interface the consumer implements. So it would involve:</p>

<p>    1. Leave `IMetricsConsumer` as-is, but deprecated.<br/>
    2. Introduce a new interface for aggregated metrics (e.g. `IAggregateMetricsConsumer`)<br/>
    3. At runtime, look at which interface the registered consumer implements, and alter the logic based on that (rather than a configuration switch)</p>

<p>    With that we would have both backward compatibility and a new presumably cleaner API for aggregated metrics. The internals my not be pretty, but that would be invisible to the user.</p>

<p>    From the discussion thread you pointed out, there seems to be consensus around such an approach:</p>

<p>    Quoting @revans2:</p>

<p>    > On 1.x we add in a new API for both reporting metrics and collecting in parallel to the old API.  We leave IMetric and IMetricsConsumer in place but deprecated.  As we move internal metrics over from the old interface to the new one, we either keep versions of the old ones in place or we provide a translation shim going from the new to the old. </p>

<p>    Quoting @abhishekagarwal87:</p>

<p>    > Sounds good. Having two separate metric reporters may be confusing but it<br/>
    is better than breaking the client code.</p>

<p>    Quoting you:</p>

<p>    > Personally I'm also in favor of maintaining old API (but deprecated) and<br/>
    adding new API. It's ideal way, and that's what many projects are trying to do, and so on<br/>
    the other project I'm also maintaining.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @ptgoetz <br/>
    That consensus was made with context `breaking change for 1.x`. Sure if we introduce breaking change in 1.x we need to support the old one. But if we go on introducing breaking change only for next major version, I don't think we must follow our consensus.</p>

<p>    We even could introduce completely different way like JStorm. In JStorm metrics are sent to Nimbus, and plugin to external storage is also there, so it might not be topology specific. In this case we could not provide the translation shim, but I don't think it matters if we introduce the change for next major version.</p>

<p>    When we talk about new API I'd like to focus the new one, not considering how to support the old one. For me, this change is more suited to the old one. There's another level to aggregate, component level, and it might be more preferred thing. Why I take worker level aggregation instead of component level aggregation is that it's only possible way to achieve with not breaking current design. That's why we should not consider about old one while designing new one.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72779215" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72779215</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java &#8212;<br/>
    @@ -436,7 +414,9 @@ public static void addMetricComponents(Map conf, StormTopology topology) {<br/>
 public static void addSystemComponents(Map conf, StormTopology topology) {<br/>
     Map<String, StreamInfo> outputStreams = new HashMap<String, StreamInfo>();<br/>
     outputStreams.put(Constants.SYSTEM_TICK_STREAM_ID, Thrift.outputFields(Arrays.asList("rate_secs")));<br/>
    +outputStreams.put(Constants.METRICS_STREAM_ID, Thrift.directOutputFields(Arrays.asList("task-info", "data-points")));<br/>
    &#8212; End diff &#8211;</p>

<p>    is it needed?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72779300" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72779300</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/metric/SystemBolt.java &#8212;<br/>
    @@ -156,7 +219,103 @@ private void registerMetrics(TopologyContext context, Map<String, String> metric</p>

<p> @Override<br/>
 public void execute(Tuple input) {</p>
<ul class="alternate" type="square">
	<li>throw new RuntimeException("Non-system tuples should never be sent to __system bolt.");
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    can we check that the incoming stream is METRICS_STREAM and throw same exception if its not?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72780835" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72780835</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/metric/SystemBolt.java &#8212;<br/>
    @@ -156,7 +219,103 @@ private void registerMetrics(TopologyContext context, Map<String, String> metric</p>

<p> @Override<br/>
 public void execute(Tuple input) {</p>
<ul class="alternate" type="square">
	<li>throw new RuntimeException("Non-system tuples should never be sent to __system bolt.");<br/>
    +IMetricsConsumer.TaskInfo taskInfo = (IMetricsConsumer.TaskInfo) input.getValue(0);<br/>
    +Collection<IMetricsConsumer.DataPoint> dataPoints = (Collection) input.getValue(1);<br/>
    +Collection<IMetricsConsumer.DataPoint> expandedDataPoints = expander.expandDataPoints(dataPoints);<br/>
    +<br/>
    +if (aggregateMode) 
{
    +    handleMetricTupleInAggregateMode(taskInfo, expandedDataPoints);
    +}
<p> else </p>
{
    +    collector.emit(Constants.METRICS_AGGREGATE_STREAM_ID, new Values(taskInfo, expandedDataPoints));
    +}
<p>    +    }<br/>
    +<br/>
    +    private void handleMetricTupleInAggregateMode(IMetricsConsumer.TaskInfo taskInfo, Collection<IMetricsConsumer.DataPoint> expandedDataPoints) {<br/>
    +Map<Integer, TaskInfoToDataPointsPair> taskToMetricTupleMap = intervalToTaskToMetricTupleMap.get(taskInfo.updateIntervalSecs);<br/>
    +if (taskToMetricTupleMap == null) </p>
{
    +    taskToMetricTupleMap = new HashMap<>();
    +    intervalToTaskToMetricTupleMap.put(taskInfo.updateIntervalSecs, taskToMetricTupleMap);
    +}
<p>    +<br/>
    +taskToMetricTupleMap.put(taskInfo.srcTaskId, new TaskInfoToDataPointsPair(taskInfo, expandedDataPoints));</p>
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    I didn't get this. It overrides the older information with newer one. what if the data point for same task is received before 'isOKtoAggregateMetricsTuples` is true. </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user abhishekagarwal87 commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @HeartSaVioR - Thanks for putting in efforts on improving metrics.  The approach of passing metrics via system bolt looks agreeable to me. Though I would still prefer "Two types of metric consumers approach". <br/>
    Secondly regarding aggregation, component level aggregation is much more preferred. That is what we do in external systems. Simply for the reason that  `total number of messages processed by component A` is more useful and actionable information than `total number of messages processed on this machine and port`. If the component information is being let go during aggregation, I don't think it will be used much. At least I wouldn't in our production system. </p>

<p>    Load among tasks is usually homogenous and uniform (law of large numbers) but same is not true for different components which execute different code altogether. I think just aggregating per worker+component combination would also reduce the amount of metrics being emitted and still keep the metrics useful.</p>

<p>    Thirdly, the assumption that a metric being declared with 60 second update interval will actually come in 60 second intervals is not correct. They can come with 120 seconds apart or they can also come 5 seconds apart (depending on how fast or slow the task is). </p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72865762" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72865762</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/metric/SystemBolt.java &#8212;<br/>
    @@ -156,7 +219,103 @@ private void registerMetrics(TopologyContext context, Map<String, String> metric</p>

<p> @Override<br/>
 public void execute(Tuple input) {</p>
<ul class="alternate" type="square">
	<li>throw new RuntimeException("Non-system tuples should never be sent to __system bolt.");<br/>
    +IMetricsConsumer.TaskInfo taskInfo = (IMetricsConsumer.TaskInfo) input.getValue(0);<br/>
    +Collection<IMetricsConsumer.DataPoint> dataPoints = (Collection) input.getValue(1);<br/>
    +Collection<IMetricsConsumer.DataPoint> expandedDataPoints = expander.expandDataPoints(dataPoints);<br/>
    +<br/>
    +if (aggregateMode) 
{
    +    handleMetricTupleInAggregateMode(taskInfo, expandedDataPoints);
    +}
<p> else </p>
{
    +    collector.emit(Constants.METRICS_AGGREGATE_STREAM_ID, new Values(taskInfo, expandedDataPoints));
    +}
<p>    +    }<br/>
    +<br/>
    +    private void handleMetricTupleInAggregateMode(IMetricsConsumer.TaskInfo taskInfo, Collection<IMetricsConsumer.DataPoint> expandedDataPoints) {<br/>
    +Map<Integer, TaskInfoToDataPointsPair> taskToMetricTupleMap = intervalToTaskToMetricTupleMap.get(taskInfo.updateIntervalSecs);<br/>
    +if (taskToMetricTupleMap == null) </p>
{
    +    taskToMetricTupleMap = new HashMap<>();
    +    intervalToTaskToMetricTupleMap.put(taskInfo.updateIntervalSecs, taskToMetricTupleMap);
    +}
<p>    +<br/>
    +taskToMetricTupleMap.put(taskInfo.srcTaskId, new TaskInfoToDataPointsPair(taskInfo, expandedDataPoints));</p>
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    It is intended behavior. If the data point for same task on same interval is received and it overwrites the old value, it means old value is not aggregated in time. If we want to keep also the old value then it's more confusing that what values we want to aggregate together. </p>

<p>    This is due to design issue on current metrics feature (late metrics) and if it occurs, metric value would be flawed anyway (I mean value is not measured for given interval) so we can't take good care of them. Fixing this issue should be a one of the requirements for new metrics feature.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72865866" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72865866</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/metric/SystemBolt.java &#8212;<br/>
    @@ -156,7 +219,103 @@ private void registerMetrics(TopologyContext context, Map<String, String> metric</p>

<p> @Override<br/>
 public void execute(Tuple input) {</p>
<ul class="alternate" type="square">
	<li>throw new RuntimeException("Non-system tuples should never be sent to __system bolt.");
	<ul class="alternate" type="square">
		<li>
		<ul class="alternate" type="square">
			<li>End diff &#8211;</li>
		</ul>
		</li>
	</ul>
	</li>
</ul>


<p>    Makes sense. I'll address this.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72866374" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72866374</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java &#8212;<br/>
    @@ -436,7 +414,9 @@ public static void addMetricComponents(Map conf, StormTopology topology) {<br/>
 public static void addSystemComponents(Map conf, StormTopology topology) {<br/>
     Map<String, StreamInfo> outputStreams = new HashMap<String, StreamInfo>();<br/>
     outputStreams.put(Constants.SYSTEM_TICK_STREAM_ID, Thrift.outputFields(Arrays.asList("rate_secs")));<br/>
    +outputStreams.put(Constants.METRICS_STREAM_ID, Thrift.directOutputFields(Arrays.asList("task-info", "data-points")));<br/>
    &#8212; End diff &#8211;</p>

<p>    Yes we send the metrics tick tuple to each task with source stream to that, so it's needed.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on a diff in the pull request:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595#discussion_r72866761" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595#discussion_r72866761</a></p>

<p>    &#8212; Diff: storm-core/src/jvm/org/apache/storm/daemon/StormCommon.java &#8212;<br/>
    @@ -436,7 +414,9 @@ public static void addMetricComponents(Map conf, StormTopology topology) {<br/>
 public static void addSystemComponents(Map conf, StormTopology topology) {<br/>
     Map<String, StreamInfo> outputStreams = new HashMap<String, StreamInfo>();<br/>
     outputStreams.put(Constants.SYSTEM_TICK_STREAM_ID, Thrift.outputFields(Arrays.asList("rate_secs")));<br/>
    +outputStreams.put(Constants.METRICS_STREAM_ID, Thrift.directOutputFields(Arrays.asList("task-info", "data-points")));<br/>
    &#8212; End diff &#8211;</p>

<p>    Oh sorry I've confused. We send the metrics tuple from each task with fake source stream which is `SYSTEM_TASK_ID METRICS_STREAM_ID`, so it's needed. Please see the change of `metric-tick`.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @abhishekagarwal87 <br/>
    1. I think new type of metrics consumer should refer the new metrics feature which addresses our concerns on current metrics. Aggregation is just a one of them so I don't want to introduce new type of metrics consumer without consideration of other things. It doesn't mean I started working for the new metrics feature, just like a hotfix for current metrics feature.</p>

<p>    2. Worker+component combination is what this PR is trying to achieve. `convertMetricsTupleMapKeyedByTaskInfo` is for that. You can see that I only change the task id and timestamp in TaskInfo, and use TaskInfo to aggregate the values, which means metrics from different components are not aggregated to same list.<br/>
    I also don't think aggregated metric value is useful if they are aggregated from various components.</p>

<p>    3. Yes I'm aware of it, and that's why I introduce `eviction` of the metric points.<br/>
    Btw, in that care, metric value would be flawed anyway. It is already out of control so we should focus how to not making this situation instead of how to correct the value.<br/>
    And triggering (sending metrics tick tuple to tasks) can't guarantee this.</p>

<p>    As we were talking about metrics from user@ and dev@ list, there're a lot of things to correct from current metrics feature. Things are coupled each other so this is also not easy to fix them one by one. Let's move out to the new metrics feature rather than saying them with current metrics.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @ptgoetz @harshach @abhishekagarwal87 <br/>
    I just had the limitations of current metrics feature only in my mind, but it would be better to write it down and share to persuade the need for new metrics feature.</p>

<p>    I didn't grant the wiki access so I just use google doc. Please let me know if we want to have this in Apache infra. I can move this out to wiki, or just post to dev@ mailing list.</p>

<p>    <a href="https://docs.google.com/document/d/11H3STXr2QTnfOqisgOnvNt8TjXS_cSqb86VM6vdAo8s/edit?usp=sharing" class="external-link" target="_blank" rel="nofollow noopener">https://docs.google.com/document/d/11H3STXr2QTnfOqisgOnvNt8TjXS_cSqb86VM6vdAo8s/edit?usp=sharing</a></p>

<p>    @unsleepy22 AFAIK you worked on the change of JStorm metrics. Could you check the doc and confirm that JStorm metrics feature addressed the limitations of Storm metrics feature?</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @HeartSaVioR It will be good to have confluent wiki setup like other projects in apache do. We can have proposal setup there and go through the process of discussion in mailing list. I would start on user list so that we can hear from any existing or future user opinions as well.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @harshach <br/>
    The wish-list contains ideas/opinions in discussions on user@ and also dev@. </p>

<p>    <a href="http://mail-archives.apache.org/mod_mbox/storm-user/201604.mbox/%3CCAF5108iwhGgae7Z2AKgMVNP4PWwV52GVc7Ex5h2oQ7fAAPX2PA%40mail.gmail.com%3E" class="external-link" target="_blank" rel="nofollow noopener">http://mail-archives.apache.org/mod_mbox/storm-user/201604.mbox/%3CCAF5108iwhGgae7Z2AKgMVNP4PWwV52GVc7Ex5h2oQ7fAAPX2PA%40mail.gmail.com%3E</a></p>

<p>    I've initiated discussions several times for this, so I'm sure what @abhishekagarwal87 sorted out is enough set of wish-list for new metrics feature.</p>

<p>    Btw, we didn't use wiki for such purpose, so it might be better to talk about the format, the process, and so on. If we just want to have a page without specific format, I'll move this to wiki page once I granted write access to wiki.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    OK. I break down my doc to two pieces and posted to wiki page.</p>

<ul class="alternate" type="square">
	<li><a href="https://cwiki.apache.org/confluence/display/STORM/Limitations+of+current+metrics+feature" class="external-link" target="_blank" rel="nofollow noopener">https://cwiki.apache.org/confluence/display/STORM/Limitations+of+current+metrics+feature</a></li>
	<li><a href="https://cwiki.apache.org/confluence/display/STORM/Wishlist+for+new+metrics+feature" class="external-link" target="_blank" rel="nofollow noopener">https://cwiki.apache.org/confluence/display/STORM/Wishlist+for+new+metrics+feature</a></li>
</ul>


<p>    Since I'm not working on new metrics feature, I didn't put them into proposal / design doc category. I think we will adopt or address many places including metrics while evaluating JStorm features from Phase 2, so would like to wait for that.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user harshach commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    @HeartSaVioR I would like to see this merged into 1.x-branch. Given suggestions here can you make the changes , introducing IAggregatedMetrics interface and leave the existing the interface as it is.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=githubbot">githubbot</a>:</i>
<p>Github user HeartSaVioR commented on the issue:</p>

<p>    <a href="https://github.com/apache/storm/pull/1595" class="external-link" target="_blank" rel="nofollow noopener">https://github.com/apache/storm/pull/1595</a></p>

<p>    We shouldn't make the change only available for 1.x, especially this is only effective when ecosystem (I mean metrics consumer implementators) should follow up. If we renew metrics feature to JStorm thing we might even don't have a chance to deprecate interfaces before removing them for 2.0.<br/>
    As I stated to wiki page, what we need to change is metrics feature itself.<br/>
    Let's wait for phase 2 to have a chance to renew, or someone claims this should be merged in 1.x branch.</p>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Closing in favor of <a href="https://issues.apache.org/jira/browse/STORM-2153" title="New Metrics Reporting API" class="issue-link" data-issue-key="STORM-2153"><del>STORM-2153</del></a>.</p>
